packages/shared/ReactMutableSource.js

```diff
@@ -0,0 +1,134 @@
+/**
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ *
+ * @flow
+ */
+
+import type {ExpirationTime} from 'react-reconciler/src/ReactFiberExpirationTime';
+import type {FiberRoot} from 'react-reconciler/src/ReactFiberRoot';
+
+// Mutable source version can be anything (e.g. number, string, immutable data structure)
+// so long as it changes every time any part of the source changes.
+export type Version = $NonMaybeType<mixed>;
+
+// Tracks expiration time for all mutable sources with pending updates.
+// Used to determine if a source is safe to read during updates.
+// If there are no entries in this map for a given source,
+// or if the current render’s expiration time is ≤ this value,
+// it is safe to read from the source without tearing.
+export type MutableSourcePendingUpdateMap = Map<
+  MutableSource<any>,
+  ExpirationTime,
+>;
+
+type MutableSourceConfig = {|
+  getVersion: () => Version,
+|};
+
+export type MutableSource<Source: $NonMaybeType<mixed>> = {|
+  _source: Source,
+
+  _getVersion: () => Version,
+
+  // Tracks the version of this source at the time it was most recently read.
+  // Used to determine if a source is safe to read from before it has been subscribed to.
+  // Version number is only used during mount,
+  // since the mechanism for determining safety after subscription is expiration time.
+  //
+  // As a workaround to support multiple concurrent renderers,
+  // we categorize some renderers as primary and others as secondary.
+  // We only expect there to be two concurrent renderers at most:
+  // React Native (primary) and Fabric (secondary);
+  // React DOM (primary) and React ART (secondary).
+  // Secondary renderers store their context values on separate fields.
+  // We use the same approach for Context.
+  _workInProgressVersionPrimary: null | Version,
+  _workInProgressVersionSecondary: null | Version,
+|};
+
+export type MutableSourceHookConfig<Source: $NonMaybeType<mixed>, Snapshot> = {|
+  getSnapshot: (source: Source) => Snapshot,
+  subscribe: (source: Source, callback: Function) => () => void,
+|};
+
+// Work in progress version numbers only apply to a single render,
+// and should be reset before starting a new render.
+// This tracks which mutable sources need to be reset after a render.
+let workInProgressPrimarySources: Array<MutableSource<any>> = [];
+let workInProgressSecondarySources: Array<MutableSource<any>> = [];
+
+export function createMutableSource<Source: $NonMaybeType<mixed>>(
```

Comment:
The use of `shared` is a last resort and not best practice. In fact, half of the files in there shouldn't be there. Including FeatureFlags that are starting to cause problems.

Even the Symbols one is causing problems because the code size duplication is quite significant. We were talking about moving them to be exports from Isomorphic instead. So I think I might do a pass trying to get rid of as much as possible from `/shared/`.

This file certainly shouldn't be in shared since `createMutableSource` and the type signatures should just move there. In terms of many files, I think we should start moving everything in isomorphic to the same file to be more careful about growing that file. It's unnecessarily big for things that can't be made lazy.

Reply:
Yeah, already moved (just not pushed yet)